### PR TITLE
memory/gmap: add test for shadow memory counters

### PR DESCRIPTION
**Pull request by smitterl**
KVM records shadow memory events for a nested guest. Confirm the counters go up.

### CODE REVIEW

**Comment by smitterl** [comment]
```
JOB ID     : e33efa43b7334838e06ee6c5638568d669825e1b
JOB LOG    : /var/lib/avocado/job-results/job-2023-11-20T08.41-e33efa4/job.log
 (1/1) type_specific.io-github-autotest-libvirt.gmap.l3_shadow_table_counters: PASS (25.72 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-11-20T08.41-e33efa4/results.html
JOB TIME   : 26.12 s
```

**Comment by smitterl** [comment]
@fbq815 Please can you help review?

**Comment by fbq815** [comment]
I have a question: is l2_guest in this code still a level 1 guest?

**Comment by smitterl** [comment]
> I have a question: is l2_guest in this code still a level 1 guest? Others LGTM

It is a level 1 kvm guest but L2 guest here because an LPAR is L1 as it's also virtualized.

**Comment by fbq815** [comment]
> > I have a question: is l2_guest in this code still a level 1 guest? Others LGTM
> 
> It is a level 1 kvm guest but L2 guest here because an LPAR is L1 as it's also virtualized.

if I understand correctly, I prefer we assume the virtual machine as L1 in confgiure file case you also name a functiona as "l1_counters_go_up". There would be some mis-understanding for others? How about we add comment like "we assume the test scenario as virtual machine which is a L2 guest in LPAR" and we only have L1 and L2 as the name?

**Comment by smitterl** [comment]
> I have a question: is l2_guest in this code still a level 1 guest? Others LGTM

Depends on the language, a mainframe LPAR is a virtualization L1, therefore a KVM guest on the LPAR is an L2. You could count the first KVM guest as L1 but that's not how it's usually counted on IBM side IIUC.

**Comment by smitterl** [comment]
> > > I have a question: is l2_guest in this code still a level 1 guest? Others LGTM
> > 
> > 
> > It is a level 1 kvm guest but L2 guest here because an LPAR is L1 as it's also virtualized.
> 
> if I understand correctly, I prefer we assume the virtual machine as L1 in confgiure file case you also name a functiona as "l1_counters_go_up". There would be some mis-understanding for others? How about we add comment like "we assume the test scenario as virtual machine which is a L2 guest in LPAR" and we only have L1 and L2 as the name?

I get your point. I updated the `run` function's comment to clarify. Please check.

**Review by fbq815 — APPROVED** [review]

File: libvirt/tests/src/memory/gmap.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        vm.start()
+        cleanup_actions.append(lambda: vm.destroy())
+        session = vm.wait_for_login()
+        cleanup_actions.append(lambda: session.close())
+        start_l3(session, target_tag)
+        if not l1_counters_go_up():
+            test.fail("Gmap counters didn't go up.")
+    finally:
+        cleanup_actions.reverse()
+        for a in cleanup_actions:
```
[the remark this answers is not in the document]
Comment (smitterl) [reply libvirt/tests/src/memory/gmap.py:141]: Here the scope is really small just this and the next line so I think the name is ok. However, I updated it to 'action'.

File: libvirt/tests/src/memory/gmap.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    vmxml.mb = mb
+    vmxml.sync()
+    fs = create_fs_xml({
+        "accessmode": "passthrough",
+        "source": {
+            "dir": parent(get_source_disk_path(vmxml))
+            },
+        "target": {"dir": target_tag},
+        "binary": {"flock": "off", "lock_posix": "off"}
+        })
```
[the remark this answers is not in the document]
Comment (smitterl) [reply libvirt/tests/src/memory/gmap.py:60]: Got it. Thanks for the example. Updated.

File: libvirt/tests/src/memory/gmap.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (smitterl) [reply libvirt/tests/src/memory/gmap.py]: @dzhengfy I addressed all your comments, please have a look, thank you!

[the remark this answers is not in the document]
Comment (smitterl) [reply libvirt/tests/src/memory/gmap.py]:
```
JOB ID     : 064246135562db85db932c63662d75fb237c078a                                                                                                                                                               
JOB LOG    : /var/lib/avocado/job-results/job-2024-02-19T05.51-0642461/job.log                                                                                                                                      
 (1/1) type_specific.io-github-autotest-libvirt.gmap.l3_shadow_table_counters: PASS (27.17 s)                                                                                                                       
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0                                                                                                                                   
JOB HTML   : /var/lib/avocado/job-results/job-2024-02-19T05.51-0642461/results.html                                                                                                                                 
JOB TIME   : 27.49 s    
```

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**